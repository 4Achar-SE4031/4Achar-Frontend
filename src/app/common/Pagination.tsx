import React from "react";
import { Pagination as MuiPagination, PaginationItem } from "@mui/material";

const Pagination: React.FC<{
  count: number;
  page: number;
  onChange: (event: React.ChangeEvent<unknown>, value: number) => void;
}> = ({ count, page, onChange }) => {
  const toPersianDigits = (num: number): string => {
    const persianDigits = "۰۱۲۳۴۵۶۷۸۹";
    return num
      .toString()
      .replace(/\d/g, (digit) => persianDigits[parseInt(digit, 10)]);
  };

  return (
    <MuiPagination
      count={count}
      page={page}
      onChange={onChange}
      renderItem={(item) => (
        <PaginationItem
          {...item}
          // Convert page numbers to Persian if the item type is 'page'
          page={item.page !== null ? toPersianDigits(item.page) : undefined}
        />
      )}
    />
  );
};

export default Pagination;
