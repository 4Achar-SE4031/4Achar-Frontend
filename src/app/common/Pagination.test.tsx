// Pagination.test.tsx
import React from 'react';
import { describe, it, expect, vi, beforeEach } from 'vitest';
import { render, screen, fireEvent } from '@testing-library/react';
import Pagination from './Pagination';

describe('Pagination Component', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  /**
   * Helper function to convert numbers to Persian digits.
   * This mirrors the `toPersianDigits` function in the component.
   */
  const toPersianDigits = (num: number): string => {
    const persianDigits = "۰۱۲۳۴۵۶۷۸۹";
    return num
      .toString()
      .replace(/\d/g, (digit) => persianDigits[parseInt(digit, 10)]);
  };

  
  it('calls onChange with the correct page number when a page button is clicked', () => {
    const count = 5;
    const currentPage = 2;
    const onChange = vi.fn();

    render(<Pagination count={count} page={currentPage} onChange={onChange} />);

    const targetPage = 4;
    const targetButton = screen.getByTestId(`pagination-button-${targetPage}`);
    fireEvent.click(targetButton);

    expect(onChange).toHaveBeenCalledTimes(1);
    expect(onChange).toHaveBeenCalledWith(expect.any(Object), targetPage);
  });

  it('calls onChange with the correct value when next button is clicked', () => {
    const count = 5;
    const currentPage = 3;
    const onChange = vi.fn();

    render(<Pagination count={count} page={currentPage} onChange={onChange} />);

    const nextButton = screen.getByTestId('pagination-button-next');
    fireEvent.click(nextButton);

    expect(onChange).toHaveBeenCalledTimes(1);
    expect(onChange).toHaveBeenCalledWith(expect.any(Object), currentPage + 1);
  });

  

  



  it('does not render first and last buttons when count is small', () => {
    const count = 4;
    const currentPage = 2;
    const onChange = vi.fn();

    render(<Pagination count={count} page={currentPage} onChange={onChange} />);

    // Depending on your implementation, first and last buttons might not be rendered
    // Adjust the assertions accordingly
    const firstButton = screen.queryByTestId('pagination-button-first');
    const lastButton = screen.queryByTestId('pagination-button-last');

    expect(firstButton).not.toBeInTheDocument();
    expect(lastButton).not.toBeInTheDocument();
  });


  it('matches the snapshot', () => {
    const count = 5;
    const currentPage = 3;
    const onChange = vi.fn();

    const { asFragment } = render(<Pagination count={count} page={currentPage} onChange={onChange} />);
    expect(asFragment()).toMatchSnapshot();
  });

  /**
   * Accessibility Test
   */


  /**
   * Test to ensure that Persian digits are correctly displayed
   */
  it('displays page numbers in Persian digits', () => {
    const count = 3;
    const currentPage = 2;
    const onChange = vi.fn();

    render(<Pagination count={count} page={currentPage} onChange={onChange} />);

    for (let i = 1; i <= count; i++) {
      const persianDigit = toPersianDigits(i);
      const pageButton = screen.getByTestId(`pagination-button-${i}`);
      expect(pageButton).toHaveTextContent(persianDigit);
    }
  });
});
