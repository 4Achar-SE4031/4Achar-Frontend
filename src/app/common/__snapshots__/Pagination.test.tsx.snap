// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Pagination Component > matches the snapshot 1`] = `
<DocumentFragment>
  <nav
    aria-label="pagination navigation"
    class="MuiPagination-root MuiPagination-text css-1xra8g6-MuiPagination-root"
  >
    <ul
      class="MuiPagination-ul css-1c5o7vy-MuiPagination-ul"
    >
      <li>
        <button
          aria-label="Go to previous page"
          class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-circular MuiPaginationItem-previousNext css-1l5xwdx-MuiButtonBase-root-MuiPaginationItem-root"
          data-testid="pagination-button-previous"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiPaginationItem-icon css-8uzxat-MuiSvgIcon-root-MuiPaginationItem-icon"
            data-testid="NavigateBeforeIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
            />
          </svg>
        </button>
      </li>
      <li>
        <button
          aria-label="Go to page 1"
          class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-circular MuiPaginationItem-page css-1l5xwdx-MuiButtonBase-root-MuiPaginationItem-root"
          data-testid="pagination-button-1"
          tabindex="0"
          type="button"
        >
          ۱
        </button>
      </li>
      <li>
        <button
          aria-label="Go to page 2"
          class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-circular MuiPaginationItem-page css-1l5xwdx-MuiButtonBase-root-MuiPaginationItem-root"
          data-testid="pagination-button-2"
          tabindex="0"
          type="button"
        >
          ۲
        </button>
      </li>
      <li>
        <button
          aria-current="true"
          aria-label="page 3"
          class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-circular Mui-selected MuiPaginationItem-page css-1l5xwdx-MuiButtonBase-root-MuiPaginationItem-root"
          data-testid="pagination-button-3"
          tabindex="0"
          type="button"
        >
          ۳
        </button>
      </li>
      <li>
        <button
          aria-label="Go to page 4"
          class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-circular MuiPaginationItem-page css-1l5xwdx-MuiButtonBase-root-MuiPaginationItem-root"
          data-testid="pagination-button-4"
          tabindex="0"
          type="button"
        >
          ۴
        </button>
      </li>
      <li>
        <button
          aria-label="Go to page 5"
          class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-circular MuiPaginationItem-page css-1l5xwdx-MuiButtonBase-root-MuiPaginationItem-root"
          data-testid="pagination-button-5"
          tabindex="0"
          type="button"
        >
          ۵
        </button>
      </li>
      <li>
        <button
          aria-label="Go to next page"
          class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-circular MuiPaginationItem-previousNext css-1l5xwdx-MuiButtonBase-root-MuiPaginationItem-root"
          data-testid="pagination-button-next"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiPaginationItem-icon css-8uzxat-MuiSvgIcon-root-MuiPaginationItem-icon"
            data-testid="NavigateNextIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
            />
          </svg>
        </button>
      </li>
    </ul>
  </nav>
</DocumentFragment>
`;
